src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,374 @@
+        auto head = _buffer.begin();
+        auto head_buf = std::move(*head).release();
+        _buffer.pop_front();
+        vlog(
+          _ctxlog.debug,
+          "get produced {} bytes, {} bytes buffered",
+          head_buf.size(),
+          _buffer.size_bytes());
+        co_return head_buf;
+    }
+    ss::future<ss::temporary_buffer<char>> skip(uint64_t n) override {
+        ss::temporary_buffer<char> empty;
+        co_return empty;
+    }
+    ss::future<> close() override {
+        co_await std::move(_reader).release()->finally();
+    }
+
+private:
+    model::record_batch_reader _reader;
+    iobuf _buffer;
+    ssize_t _max_bytes;
+    size_t _num_batches{};
+    model::timeout_clock::time_point _deadline;
+    inclusive_offset_range _range;
+    ss::abort_source _as;
+    retry_chain_node _rtc;
+    retry_chain_logger _ctxlog;
+};
+
+ss::input_stream<char> make_reader_input_stream(
+  model::ntp ntp,
+  model::record_batch_reader r,
+  size_t read_buffer_size,
+  inclusive_offset_range range,
+  model::timeout_clock::duration timeout) {
+    auto deadline = model::timeout_clock::now() + timeout;
+    auto ds = std::make_unique<reader_ds>(
+      ntp, std::move(r), read_buffer_size, range, deadline);
+    ss::input_stream<char> s(ss::data_source(std::move(ds)));
+    return s;
+}
+
+segment_upload::segment_upload(
+  ss::lw_shared_ptr<cluster::partition> part,
+  size_t read_buffer_size,
+  ss::scheduling_group sg)
+  : _ntp(part->get_ntp_config().ntp())
+  , _part(part)
+  , _rd_buffer_size(read_buffer_size)
+  , _sg(sg)
+  , _rtc(_as)
+  , _ctxlog(archival_log, _rtc, _ntp.path()) {}
+
+void segment_upload::throw_if_not_initialized(std::string_view caller) const {
+    if (_stream.has_value() == false) {
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format,
+          "Segment upload is not initialized {} to {}",
+          _ntp,
+          caller));
+    }
+}
+
+ss::future<result<std::unique_ptr<segment_upload>>>
+segment_upload::make_segment_upload(
+  ss::lw_shared_ptr<cluster::partition> part,
+  inclusive_offset_range range,
+  size_t read_buffer_size,
+  ss::scheduling_group sg,
+  model::timeout_clock::time_point deadline) {
+    std::unique_ptr<segment_upload> upl(
+      new segment_upload(part, read_buffer_size, sg));
+
+    auto res = co_await upl->initialize(range, deadline);
+    if (res.has_failure()) {
+        co_return res.as_failure();
+    }
+    co_return std::move(upl);
+}
+
+ss::future<result<std::unique_ptr<segment_upload>>>
+segment_upload::make_segment_upload(
+  ss::lw_shared_ptr<cluster::partition> part,
+  size_limited_offset_range range,
+  size_t read_buffer_size,
+  ss::scheduling_group sg,
+  model::timeout_clock::time_point deadline) {
+    std::unique_ptr<segment_upload> upl(
+      new segment_upload(part, read_buffer_size, sg));
+
+    auto res = co_await upl->initialize(range, deadline);
+    if (res.has_failure()) {
+        co_return res.as_failure();
+    }
```

Comment:
nit: now that there is a public static initializer and no public constructor, WDYT about adding the stream as an argument to the `segment_upload` constructor. Perhaps we can run the bits in `initialize()` first to construct a stream, and then pass it in as a member to the `segment_upload`. That way there's never a time where `segment_upload` isn't ready for action, and we can remove the calls to `throw_if_not_initialized()`